### PR TITLE
sync: refactored `PollSender<T>` to fix a subtly broken `Sink<T>` implementation

### PR DESCRIPTION
**Pull request by tobz**
## Motivation

The existing `PollSender<T>` implementation uses a simplified design that treats `poll_send_done` as both `poll_ready` and `poll_flush`.  However, it only ever actually does any flushing work i.e. driving the future which sends into the underlying channel. When used in `poll_ready`, it simply returns that the sender is always ready.

This leads to a subtle bug where callers could inadvertently use something like `sink.send_all(..).await` which leads to a panic. Combinators like `send_all` believe they can continue to call `start_send` so long as `poll_ready` returns `true` when called directly before.  However, in the current design, if the underlying channel was full, the next call to `start_send` -- after `poll_ready` claimed the sender was ready -- could hit a scenario where the pending send was not yet complete, which would leave `self.is_sending` as `true`, causing the next call to `start_send` to hit the check for if `self.is_sending` is `true`, thus leading to a panic. 

## Solution

This PR explores a refactored design where instead of driving a channel send future, we drive a future for reserving a permit for sending to the channel.  This moves the ordering of logic from `poll_send_done`/`start_send`/`poll_send_done` to simply `poll_reserve`/`start_send`.

This allows `poll_ready` to actually be representative of whether or not an item can be sent into the channel.  Additionally, through the state machine approach, we can eliminate most clones of the `Sender` itself by recapturing the underlying sender from the permit.

This does change a few methods and some naming so it would be a breaking change, although I think it's worth it for the ability to correctly provide the `Sink<T>` implementation.

Signed-off-by: Toby Lawrence <toby@nuclearfurnace.com>

### CODE REVIEW

File: tokio-util/src/sync/mpsc.rs [hunk 1]
```diff
@@ -1,220 +1,274 @@
...
+    /// 
+    /// If the channel is closed, an error will be returned.  This is a permanent state.
+    /// 
+    /// # Panics
+    /// 
+    /// If `poll_reserve` was not successfully called prior to calling `start_send`, then this method
+    /// will panic.
+    pub fn start_send(&mut self, value: T) -> Result<(), PollSendError<T>> {
+        let (result, next_state) = match self.take_state() {
+            State::Idle(_) | State::Acquiring => panic!("`start_send` called without first calling `poll_ready`"),
```
Comment (Darksonn) [inline tokio-util/src/sync/mpsc.rs:139]: We could make the choice to have `start_send` succeed in this case, attempting to use the `send` method instead?

Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:139]: The only viable approach I currently see to make this method more resilient is that we could _attempt_ a send via `Sender::try_send`.  We'd then be able to change the language in the documentation to something more like:

> If `poll_reserve` is called prior to calling `start_send`, and returns `Poll::Ready(Ok(()))`, then the call to `start_send` is guaranteed to succeed.  If it is not called prior, then a send is attempted but may or may not succeed.

The biggest issue then would be how to document the error case: `PollSendError<T>` is meant to only ever indicate that the channel is closed, but in this case, `try_send` might fail due to the channel being closed _or_ full.  It feels weird to differentiate errors based on the condition of whether or not `poll_reserve` was called since there is supposed to be a formal contract of calling `poll_reserve` _before_ `start_send`.

Long story short, I'm willing to make such a change, just curious what your thoughts are on how to best document the behavior difference depending on if `poll_reserve` was called first or not.

File: tokio-util/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+# Unreleased
+
+### Fixed
+- sync: refactored `PollSender<T>` to fix a subtly broken `Sink<T>` implementation ([#xxxx])
+
```
[the remark this answers is not in the document]
Comment (tobz) [reply tokio-util/CHANGELOG.md]: Understood, habit from another project.

Fixed in 105c02c.

File: tokio-util/src/sync/mpsc.rs [hunk 3]
```diff
@@ -1,221 +1,282 @@
...
     }
 
-    /// Access the underlying `Sender`.
+    /// Gets a reference to the underlying `Sender`.
     ///
-    /// If this method returns `None`, then the channel is closed. (But it is
-    /// not guaranteed to return `None` if the channel is closed.)
-    pub fn inner_ref(&self) -> Option<&Sender<T>> {
-        self.sender.as_deref()
+    /// If the channel is closed, `None` is returned.
```
Comment (Darksonn) [inline tokio-util/src/sync/mpsc.rs:168]: This might return `Some` even if the channel is closed.

File: tokio-util/src/sync/mpsc.rs [hunk 4]
```diff
@@ -1,221 +1,282 @@
...
 
-    /// Clone the underlying `Sender`.
+    /// Checks whether this sender is been closed.
     ///
-    /// If this method returns `None`, then the channel is closed. (But it is
-    /// not guaranteed to return `None` if the channel is closed.)
-    pub fn clone_inner(&self) -> Option<Sender<T>> {
-        self.sender.as_ref().map(|sender| (&**sender).clone())
+    /// The underlying channel that this sender was wrapping may still be open.
+    pub fn is_closed(&self) -> bool {
```
Comment (Darksonn) [inline tokio-util/src/sync/mpsc.rs:162]: You don't want to check whether the underlying channel is closed here?

File: tokio-util/src/sync/mpsc.rs [hunk 5]
```diff
@@ -1,221 +1,282 @@
...
+    /// No more messages will be able to be sent from this sender, but the underlying channel will
+    /// remain open until all senders have dropped, or until the [`Receiver`] closes the channel.
     ///
-    /// If there is a send in progress when this method is called, that send is
-    /// unaffected by this operation, and `poll_send_done` can still be called
-    /// to complete that send.
+    /// If a slot was previously reserved by calling `poll_reserve`, then a final call can be made
+    /// to `start_send` in order to consume the reserved slot.  After that, no further sends will be
+    /// possible.  If you do not intend to send another item, you can release the reserved slot back
+    /// to the underlying sender by calling [`abort_send`].
```
Comment (Darksonn) [inline tokio-util/src/sync/mpsc.rs:181]: This seems like a footgun.

File: tokio-util/src/sync/mpsc.rs (hunk 3 above)
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:168]: I guess the better comment would be "If this sender is closed, `None` is returned.`"

The implication is that we drop our references to the underlying channel sender when asked to close.

File: tokio-util/src/sync/mpsc.rs (hunk 5 above)
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:181]: Is the "footgun" aspect you're referring to specifically about how PollSender can keep a Receiver open (when PollSender is in the ReadyToSend state) even when all other Sender references are gone? I was mostly trying to follow the previous behavior where an in-flight "operation" could be finished cleanly even after closing the PollSender.

If you think we should abandon that behavior, I'm happy to make close be more forceful.

File: tokio-util/src/sync/mpsc.rs (hunk 3 above)
Comment (Darksonn) [reply tokio-util/src/sync/mpsc.rs:168]: There is a comment similar to that on the existing implementation.

Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:168]: That situation seems, potentially, very normal: `Receiver` closes itself, a bunch of `Sender`s still exist and haven't yet tried to send, or check the channel status, to figure out the channel is closed and that they can't send anything else.

If a user calls `poll_reserve` and we encounter an error reserving the permit, we mark _ourselves_ as closed, so our close state is eventually consistent with the underlying channel.

Comment (Darksonn) [reply tokio-util/src/sync/mpsc.rs:168]: I don't think there's anything wrong with how its implemented here - I just think the comment should be updated.

Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:168]: Sorry, looks like I responded to the top-level comment twice from a stale Chrome tab or something.

I'll update the comment. 👍🏻

Comment (tobz) [reply tokio-util/src/sync/mpsc.rs]: Addressed in eeb4abdf.

File: tokio-util/src/sync/mpsc.rs (hunk 4 above)
Comment (Darksonn) [reply tokio-util/src/sync/mpsc.rs:162]: I don't think the current state of this method is great.

File: tokio-util/src/sync/mpsc.rs [hunk 6]
```diff
@@ -1,221 +1,283 @@
...
-    }
+    /// # Errors
+    ///
+    /// If the channel is closed, an error will be returned.  This is a permanent state.
+    ///
+    /// # Panics
+    ///
+    /// If `poll_reserve` was not successfully called prior to calling `start_send`, then this method
+    /// will panic.
+    pub fn start_send(&mut self, value: T) -> Result<(), PollSendError<T>> {
```
Comment (Darksonn) [inline tokio-util/src/sync/mpsc.rs:139]: Since this always completes immediately, maybe the name of this method should be changed.

File: tokio-util/src/sync/mpsc.rs (hunk 4 above)
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:162]: I think most of the feedback has revolved around the lengths the PR goes to in order to attempt to preserve an acquired sending slot, so let me ask: should we actually bother trying to hold an acquired sending slot?

If we get rid of that functionality, things get much simpler.  I don't have a strong op inion on keeping that behavior, and I'm not sure that it provides more flexibility compared to simply avoiding closing the sender before you're done actually sending.

File: tokio-util/src/sync/mpsc.rs [hunk 7]
```diff
@@ -1,221 +1,283 @@
...
-        self.sender.as_deref()
+    /// If `PollSender` has been closed, `None` is returned. The underlying channel that this sender
+    /// was wrapping may still be open.
+    pub fn get_ref(&self) -> Option<&Sender<T>> {
+        self.sender.as_ref()
     }
 
-    // This operation is supported because it is required by the Sink trait.
-    /// Close this sender. No more messages can be sent from this sender.
+    /// Closes this sender without dropping it.
```
[the remark this answers is not in the document]
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:174]: Hmmmm, yeah, you're right.  The change in wording went a little too far in the other direction.  I'll fix that.

File: tokio-util/src/sync/mpsc.rs (hunk 6 above)
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:139]: Fair point, and it would also avoid the ambiguous naming situation. 👍🏻

File: tokio-util/src/sync/mpsc.rs (hunk 5 above)
Comment (Darksonn) [reply tokio-util/src/sync/mpsc.rs:182]: Regardless of the behavior of this method, I think that `Sink::poll_close` should properly close it.

As for in-flight operations, it's true that we have a similar API in Tokio, but it's on the receiver and necessary there because of races, but that race isn't present for senders, so its not as necessary.

Comment (tobz) [reply tokio-util/src/sync/mpsc.rs:182]: But what is the proper way then?  Making any subsequent call to `poll_ready`/`start_send`/`poll_flush` not do anything?

File: tokio-util/src/sync/mpsc.rs (hunk 6 above)
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs]: Addresses in 04413d6d.

File: tokio-util/src/sync/mpsc.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tobz) [reply tokio-util/src/sync/mpsc.rs]: Addressed in 04413d6d.

**Review by Darksonn — APPROVED** [review]
